python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+        # if self._getNumGoodPixels(maskedIm) >= 1:
+        disp.scale('asinh', 'zscale')
+        disp.setMaskTransparency(80)
+        disp.setMaskPlaneColor("BAD", afwDisplay.RED)
+
+        disp.setImageColormap('gray')
+        title = f"Detector: {detector.getName()[-3:]} {detector.getSerial()}"
+        title += f", Type: {imageType}, visit: {visit}"
+        disp.mtv(maskedIm, title=title)
+
+        cgUtils.overlayCcdBoxes(detector, isTrimmed=True, display=disp)
+
+    def _plotAmpHistogram(self, dataRef, exp, visit, nSigmaUsed):  # pragma: no cover
+        """
+        Make a histogram of the distribution of pixel values for each amp.
+
+        Main image data histogram plotted in blue. Edge pixels, if masked, are
+        in red. Note that masked edge pixels do not contribute to the
+        underflow and overflow numbers.
+
+        Parameters
+        ----------
+        pdfPages: `matplotlib.backends.backend_pdf.PdfPages`
+            If not None, append per-amp histograms to this object
+        """
+        fig, ax = plt.subplots(nrows=4, ncols=4, sharex='col', sharey='row', figsize=(13, 10))
+
+        detector = exp.getDetector()
+        expTime = exp.getInfo().getVisitInfo().getExposureTime()
+
+        for (amp, a) in zip(reversed(detector), ax.flatten()):
+            mi = exp.maskedImage[amp.getBBox()]
+
+            # normalise by expTime as we plot in ADU/s and don't always work with master calibs
+            mi.image.array /= expTime
+            stats = afwMath.makeStatistics(mi, afwMath.MEANCLIP | afwMath.STDEVCLIP)
+            mean, sigma = stats.getValue(afwMath.MEANCLIP), stats.getValue(afwMath.STDEVCLIP)
+
+            # Get array of pixels
+            EDGEBIT = exp.maskedImage.mask.getPlaneBitMask("EDGE")
+            imgData = mi.image.array[(mi.mask.array & EDGEBIT) == 0].flatten()
+            edgeData = mi.image.array[(mi.mask.array & EDGEBIT) != 0].flatten()
+
+            thrUpper = mean + nSigmaUsed*sigma
+            thrLower = mean - nSigmaUsed*sigma
+
+            nRight = len(imgData[imgData > thrUpper])
+            nLeft = len(imgData[imgData < thrLower])
+
+            nsig = 6
+            leftEdge = mean - nsig * nSigmaUsed*sigma
+            rightEdge = mean + nsig * nSigmaUsed*sigma
+            nbins = np.linspace(leftEdge, rightEdge, 1000)
+            ey, bin_borders, patches = a.hist(edgeData, histtype='step', bins=nbins, lw=1, edgecolor='red')
+            y, bin_borders, patches = a.hist(imgData, histtype='step', bins=nbins, lw=3, edgecolor='blue')
+
+            # Report number of entries in over-and -underflow bins, i.e. off the edges of the histogam
+            nOverflow = len(imgData[imgData > rightEdge])
+            nUnderflow = len(imgData[imgData < leftEdge])
+
+            # Put v-lines and textboxes in
+            a.axvline(thrUpper, c='k')
+            a.axvline(thrLower, c='k')
+            msg = f"{amp.getName()}\nmean:{mean: .1f}\n$\sigma$:{sigma: .1f}"  # noqa: W605
```

Comment:
Why does flake8 complain about invalid escapes on this line?

Reply:
I need to pass a `\sigma` to the matplotlib latex text generator. If I put an `r` on the front of the string then the `\n` get sent literally, which is wrong, and matplotlib seems to know what to do correctly when it's passed as-is. I messed around with some other things and it wasn't a trivial fix, so just stuck in the `noqa`, but could also probably do it by breaking the message up into multiple lines, adding the `\n`s and the `\sigma` in different places so that some strings are `r` strings, and others not. Is that worth doing? Other thoughts?

Reply:
It's telling you to do `\\sigma` there.

Reply:
🤦‍♂

Reply:
Thanks @timj!